File: Data-Structures/Matrix-Manipulation/0054_spiral_matrix.py
#matrix_manipulation

class Solution:
    def spiralOrder(self, matrix):
     #dx, dy = -dy, dx, change when out of bounds 
        rows, cols = len(matrix), len(matrix[0])
        r, c, dr, dc = 0, 0, 0, 1
        res = []

        for _ in range(rows*cols):
            res.append(matrix[r][c])
            matrix[r][c] = "." #seen 
            if not (0 <= r + dr < rows and 0 <= c + dc < cols and matrix[r+dr][c+dc] != "."):
                dr, dc = dc, -dr #change when out of bounds 
            r += dr
            c += dc
        return res

File: Data-Structures/Matrix-Manipulation/0054_spiral_matrix.c
// Practice for: 0054_spiral_matrix.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>

int* spiralOrder(int** matrix, int matrixSize, int* matrixColSize, int* returnSize){
    if(matrixSize = 0){
        *returnSize = 0;
        return NULL;
    }

    int rows = matrixSize;
    int cols = matrixColSize[0];
    int total = rows * cols;
    int* result = (int*)malloc(total * sizeof(int));

    int top = 0;
    int bottom = rows - 1;
    int left = 0;
    int right = cols - 1;

    while(*returnSize < total){
        //traverse right
        for( int i = left; i <= right && *returnSize < total; i++){
            result[(*returnSize)++] = matrix[top][i];
        }
        top++;

        //traverse down
        for (int i = top; i <= bottom && *returnSize < total; i++){
            result[(*returnSize)++] = matrix[i][right];
        }
        right--;

        //traverse left
        for (int i = right; i >= left && *returnSize < total; i++){
            result[(*returnSize)++] = matrix[bottom][i];
        }
        bottom--;

        //traverse up
        for (int i = bottom; i >= top && *returnSize < total; i++){
            result[(*returnSize)++] = matrix[i][left];
        }
        left++;
    }
    return result;
}
